proposals/semi-auto-properties.md

```diff
@@ -10,6 +10,228 @@ Two common scenarios are that you want to apply a constraint on the setter, ensu
+_ = new Derived();
+
+class Base
+{
+    public virtual bool IsActive { get; set; } = true;
+}
+
+class Derived : Base
+{
+    public override bool IsActive
+    {
+        get => base.IsActive;
+        set
+        {
+            base.IsActive = value;
+            Console.WriteLine("This will not be reached");
+        }
+    }
+}
+```
+
+### Constructor assignment
+
+As with existing auto-properties, assignment in the constructor calls the setter if it exists, and if there is no setter it falls back to directly assigning to the backing field.
+
+```cs
+class C
+{
+    public C()
+    {
+        P1 = 1; // Assigns P1's backing field directly
+        P2 = 2; // Assigns P2's backing field directly
+        P3 = 3; // Calls P3's setter
+    }
+
+    public int P1 => field;
+    public int P2 { get => field; }
+    public int P3 { get => field; set => field = value; }
+}
+```
+
+### Definite assignment in structs
+
+A semi-auto property will be treated as a regular auto property for the purposes of calculating default backing field initialization if its setter is automatically implemented, or if it does not have a setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-02.md#property-assignment-in-structs)).
+
+Default initialize a struct when calling a manually implemented semi-auto property setter, and issue a warning when doing so, like a regular property setter ([LDM decision](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-02.md#definite-assignment-of-manually-implemented-setters)).
+
+### Nullability
+
+When `{ get; }` is written as `{ get => field; }`, or `{ get; set; }` is written as `{ get => field; set => field = value; }`, a similar warning should be produced when a non-nullable property is not initialized:
+
+```cs
+class C
+{
+    // ⚠️ CS8618: Non-nullable property 'P' must contain a
+    // non-null value when exiting constructor.
+    public string P { get => field; set => field = value; }
+}
+```
+
+No warning should be produced if the property is initialized to a non-null value via constructor assignment or property initializer:
+
+```cs
+class C
+{
+    public C() { P = ""; }
+
+    public string P { get => field; set => field = value; }
+}
+```
+
+```cs
+class C
+{
+    public string P { get => field; set => field = value; } = "";
+}
+```
+
+For reference-typed fields, the `field` type should be nullable. No warning should be produced in the following example:
+
+```cs
+class C
+{
+    public string P => field ??= "test";
+}
+```
+
+### `nameof`
+
+`nameof(field)` will be disallowed. It is not like `nameof(value)`, which is the thing to use when property setters throw ArgumentException as some do in the .NET core libraries. In contrast, `nameof(field)` has no expected use cases. If it did anything, it would return the string `"field"`, consistent with how `nameof` behaves in other circumstances by returning the C# name or alias, rather than the metadata name.
```

Comment:
Think we should have a "Breaking Changes" section where items like this are discussed. This is a breaking change for anyone using `nameof(field)` today inside of a property accessor. 

Also is this an error or warning?

Reply:
Is the breaking change specific to `nameof(field)` or simply following from the meaning of `field` itself changing?

Reply:
I'd suggest an error, just like `nameof(nint)`.

Reply:
It's following from the meaning of `field` changing. Based on the spec I would expect the following to succeed in C# 12 bit result in a diagnostic for C# 13. 

```csharp
public int field;
public string Prop
{
  get => "example {nameof(field)}";
}
```

Reply:
> I'd suggest an error, just like nameof(nint).

Thought LDM reviewed this and was considering having `nameof(keyword)` work but result in a warning. Did we decide to keep the current behavior of error? If so ... may want to try and revisit that. The current behavior is adding upgrade friction that could be avoided.

Reply:
@jaredpar Last time I was in the loop my understanding was, compiler vNext warns as long as your langversion is 12 or lower, and interprets `field` as a keyword when your langversion is 13+.

Reply:
> For nameof(field) specifically, last time I was in the loop my understanding was that compiler vNext would warn as long as your langversion is 12 or lower, and would interpret field as a keyword when your langversion is 13+ with all the consequences that keywordness entails.

Doesn't that lead to a situation though where multi-targeting becomes very difficult? Lets say that you use the constant `"field"` as that will compile in both C# 12 and 13. Won't the C# 12 target sart to issue warnings / infos saying you should use `nameof(field)` instead of a constant? Or does that all get handled correctly?

These type of issues always feel to me like we're creating problems where we don't need to. Every tweak like this is just another rule that generators, code style, analyzers, has to learn and adapt to. As opposed to say allowing `nameof(...)` anything and issuing a warning (which is easy to silence) for specific keywords.

Reply:
It's a great point about unnecessary upgrade pain. I'll think of some full examples of how this will go with both `field` on its own and (the likely comparatively rarer) `nameof(field)`, upgrading through various means including manual `<LangVersion>` edits.

Reply:
@jaredpar The scenario I've thought of where the C# 12 target will issue warning or info recommending `nameof(field)` is if `field` is the name of a parameter _and_ you're passing the string to an `ArgumentException`. (<https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1507#cause>) `field` as the name of the property is another possibility, but someone would have had to already started turning off linting warnings for not using pascal casing.

However, you can't legitimately throw ArgumentException directly from an accessor body with `"field"`, since `"field"` is not the metadata name of a parameter of the accessor method. So if the linter is giving you trouble, it's pointing the way to a bigger issue worth fixing.

Considering local functions and lambdas inside an accessor body which may declare a parameter called `field` and throw such an ArgumentException, which the linter may recommend `nameof(field)` for, the LDM decision is that this parameter is shadowing the property's backing field, and so `nameof(field)` would not produce an error.

(https://github.com/dotnet/csharplang/issues/7964 confirms that `field` is only a keyword within property accessors, and together with the shadowing rule, we can be sure that the IDE will not produce warnings in C# 13 for `nameof(field)` unless it's being improperly used in the first place.)

I'll flesh out the section to clarify all this.

Reply:
Come to think of it, I don't think the nature of the break with `nameof(field)` is any different (or should be any different) than the nature of the break with `M(field);` when its meaning changes. The `nameof` error falls out of that same essential break when `field` starts meaning something different.

Reply:
@jaredpar I pushed a much smaller update, since it's out of scope for me in this PR to update the speclet to propose _any_ breaking changes at all (it's still only an open question at the end). That'll need to be done. The way I'm seeing it, the `nameof(field)` break is one and the same with the `field` break itself, and what we really need to contend with is what happens if you use `field` when multitargeting 12 and 13.